x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/SharedBytes.java

```diff
@@ -28,136 +35,141 @@
-                throw new UnsupportedOperationException();
-            }
-
-            @Override
-            public long read(ByteBuffer[] dsts, int offset, int length) throws IOException {
-                throw new UnsupportedOperationException();
-            }
+    // TODO: dry up against MLs usage of the same method
+    private static long getUsableSpace(Path path) throws IOException {
+        long freeSpaceInBytes = Environment.getFileStore(path).getUsableSpace();
 
-            @Override
-            public int write(ByteBuffer src) throws IOException {
-                throw new UnsupportedOperationException();
-            }
-
-            @Override
-            @SuppressForbidden(reason = "Use positional writes on purpose")
-            public long write(ByteBuffer[] srcs, int offset, int length) throws IOException {
-                checkOffsets(offset, length);
-                return fileChannel.write(srcs, offset, length);
-            }
-
-            @Override
-            public long position() throws IOException {
-                return fileChannel.position();
-            }
-
-            @Override
-            public FileChannel position(long newPosition) throws IOException {
-                checkOffsets(newPosition, 0);
-                return fileChannel.position(newPosition);
-            }
+        /* See: https://bugs.openjdk.java.net/browse/JDK-8162520 */
+        if (freeSpaceInBytes < 0) {
+            freeSpaceInBytes = Long.MAX_VALUE;
+        }
+        return freeSpaceInBytes;
+    }
 
-            @Override
-            public long size() throws IOException {
-                return fileChannel.size();
-            }
+    @Override
+    protected void closeInternal() {
+        try {
+            IOUtils.close(fileChannel, path == null ? null : () -> Files.deleteIfExists(path));
+        } catch (IOException e) {
+            logger.warn("Failed to clean up shared bytes file", e);
+        }
+    }
 
-            @Override
-            public FileChannel truncate(long size) throws IOException {
-                assert false;
-                throw new UnsupportedOperationException();
-            }
+    private final Map<Integer, IO> ios = ConcurrentCollections.newConcurrentMap();
 
-            @Override
-            public void force(boolean metaData) throws IOException {
-                fileChannel.force(metaData);
-            }
-
-            @Override
-            public long transferTo(long position, long count, WritableByteChannel target) throws IOException {
-                checkOffsets(position, count);
-                return fileChannel.transferTo(position, count, target);
+    IO getFileChannel(int sharedBytesPos) {
+        assert fileChannel != null;
+        return ios.compute(sharedBytesPos, (p, io) -> {
+            if (io == null || io.tryIncRef() == false) {
+                final IO newIO;
+                boolean success = false;
+                incRef();
```

Comment:
tryIncref already incremented by 1?

Reply:
This is for `SharedBytes` not `IO`. We have a ref count on the `SharedBytes` here that is incremented for each `IO` we created and one for the `IO` itself. The `IO` will decrement shared bytes once its fully released.

Reply:
ok, we should have tests for these eventually as well.